### PR TITLE
Update outer boundary implementation to support open boundary conditions

### PR DESCRIPTION
**Pull request by apcraig**
## PR checklist
- [X] Short (1 sentence) summary of your PR: 
    Update outer boundary implementation to support open boundary conditions
- [X] Developer(s): 
    apcraig
- [X] Suggest PR reviewers from list in the column to the right.
- [X] Please copy the PR test results link or provide a summary of testing completed below.
    All tests are bit-for-bit except the halochk unittest which was refactored and passes.  There is an issue in the evp1d implementation where on some compilers when the debug flag is on, the code will trap on a floating point error.  See more details below.  https://github.com/CICE-Consortium/Test-Results/wiki/cice_by_hash_forks#85e4ed8a966197ba8479928cbd93ba8e0597df1a, https://github.com/CICE-Consortium/Test-Results/wiki/cice_by_hash_forks#a8862c3bc36f31d3f5fa07d180f45e3a8e8979b6
- How much do the PR code changes differ from the unmodified code? 
    - [X] bit for bit
    - [ ] different at roundoff level
    - [ ] more substantial 
- Does this PR create or have dependencies on Icepack or any other models?
    - [ ] Yes
    - [X] No
- Does this PR update the Icepack submodule?  If so, the Icepack submodule must point to a hash on Icepack's main branch.
    - [ ] Yes
    - [X] No
- Does this PR add any new test cases?
    - [ ] Yes
    - [X] No
- Is the documentation being updated? ("Documentation" includes information on the wiki or in the .rst files from doc/source/, which are used to create the online technical docs at https://readthedocs.org/projects/cice-consortium-cice/. A test build of the technical docs will be performed as part of the PR testing.)
    - [ ] Yes
    - [X] No, does the documentation need to be updated at a later time?
        - [ ] Yes
        - [X] No 
- [X] Please document the changes in detail, including _why_ the changes are made.  This will become part of the PR commit log.

Refactor CICE to support open boundary conditions where values may be imposed on the outer boundary.  Prior to this update, the HaloUpdate was zeroing out the halo before updating the halo.  This meant any values set on the halo were lost anytime a HaloUpdate was executed.  Several other issues upstream and downstream were found as this was implemented.  In particular, the global index set on the outer boundary for open and closed boundary conditions was "special", a zero global index was used internally as a flag, and both the scatter_global and haloUpdate was covering up some issues with uninitialized data.

Update the i_glob and j_glob indexing for open and closed boundary conditions on the outer boundary.  For open, closed, and tripole (south) boundaries, now extrapolate index values on the outer boundary.  This might mean indices are less than 0 or greater than n[x,y]_global+2*nghost.  In the original implementation, these outer boundary indices were either set to some internal values so they would be set with the scatter_global method or to zero which was treated as a special value.  The zero special value feature has been removed and the scatter_global implementation has been refactored to fill all gridpoints with valid global indices and leave others unset.  The only special case remaining in the global indexing is for tripole grids on the north boundary where the j global index is set to a negative value.
    
- Refactor HaloUpdate in ice_boundary in serial and mpi to avoid zeroing out the halo.  The ew and ns directions are handled separately.  The new implementation will always zero the internal halo points.  It will also zero the outer halo if cyclic or tripole is specified or if a fillvalue is passed into the haloUpdate.  Otherwise, the outer halo is left unchanged.
  - Added ewBoundaryType and nsBoundaryType to halo datatype to keep track of the boundary conditions for each halo datatype.
  - Zero out the outer boundary during HaloUpdate only under certain conditions including cyclic or tripole bcs or if a fillValue is explicitly passed
- Add a few fillvalue=c0 arguments to HaloUpdate calls during initialization to force the outer boundary values to zero, mostly during initialization of grid data and similar cases.
- Add several HaloExtrapolate calls to initialize boundary values for grid fields.  This replaces those fields being set by the incorrect global indices and the scatter.
- Explicitly zero out all allocated variables at initialization
- Explicitly zero out restart fields before reading to be extra careful
- Explicitly zero out some arrays in the evp1d before use
- Replace the G_HTN and G_HTE initialization with a call to gather_global_ext on HTN and HTE respectively after HTN and HTE are set and the halos updated.
- Move the closed boundary condition abort to initialization in ice_domain.F90
- Refactor halochk unit test, add explicit fill tests

Additional Features
- Update ice_HaloExtrapolate to support more than 1 ghost cell
- Add support for kmt_type = none with in subroutine rectgrid for rectangular grids.  This allows a user to setup a rectangular grid with no land.
- Update the restart_ext implementation in ice_read_write.  There was a bug if restart_ext=.false. was passed, nothing would happen.  The update also allows some extra restart_ext logic in io_netcdf/ice_restart.F90 to be cleaned up.
- Carry out some code cleanup in init_domain_distribution and init_grid2
- Rename iglb/jglb to isrc/jsrc consistent in scatter_global_ext to be consistent with other scatter methods.

All tests are bit-for-bit except the halochk unittest which was refactored and passes.  There is an issue in the evp1d implementation where on some compilers when the debug flag is on, the code will trap on a floating point error.  If the floating point error isn't trapped, the cases run to completion and are bit-for-bit with the current trunk.  This suggests there is something like a 0/0 calculation being trapped in the evp1d which has no impact on the solution, is only caught by a subset of compilers, and probably requires an if test to avoid.  I spent some time trying to track down the error without any success.

### CODE REVIEW

File: cicecore/cicedyn/dynamics/ice_dyn_evp1d.F90 [hunk 1]
```diff
@@ -628,6 +627,11 @@ subroutine gather_static(G_uarear, G_dxT, G_dyT, G_Tmask)
 
      character(len=*), parameter :: subname = '(gather_static)'
 
+     G_uarear = c0
+     G_dyT = c0
```
[the remark this answers is not in the document]
Comment (apcraig) [reply cicecore/cicedyn/dynamics/ice_dyn_evp1d.F90]: This may not be necessary in practice if we can be sure that the gather_global_ext sets all gridcells in the G_ arrays.  But the cost of zeroing the arrays at initialization is also very small and probably reasonable to do.  The hope is that zeroing these arrays makes the implementation more robust.  Given we have a small technical issue in the evp1d implementation where some compilers trap a floating point error, this seems like a sensible thing to do for now.

[the remark this answers is not in the document]
Comment (apcraig) [reply cicecore/cicedyn/dynamics/ice_dyn_evp1d.F90]: A slightly bigger worry is resetting all the G_ arrays to zero in convert_1d_2d_dyn below which happens each timestep.  Again, it may improve robustness in the short term.  But at some point, we may decide it's unnecessary and want to remove that code again.

[the remark this answers is not in the document]
Comment (TillRasmussen) [reply cicecore/cicedyn/dynamics/ice_dyn_evp1d.F90]: Many of these conversions are not really needed  or only needed for history output which should be done smarter than it currently is (only convert when needed).

File: cicecore/cicedyn/infrastructure/comm/mpi/ice_gather_scatter.F90 [hunk 2]
```diff
@@ -1832,7 +1753,7 @@ subroutine scatter_global_dbl(ARRAY, ARRAY_G, src_task, dst_dist, &
    endif
 
    !-----------------------------------------------------------------
-   ! Ensure unused ghost cell values are 0
+   ! Ensure/reset ghost cell values are 0 for noupdate
```
[the remark this answers is not in the document]
Comment (apcraig) [reply cicecore/cicedyn/infrastructure/comm/mpi/ice_gather_scatter.F90]: Fixed.

**Comment by apcraig** [comment]
> Does the documentation need to be updated re open/closed boundaries for regional configurations? See section 3.1.2.5.

I meant to update that section, thanks for reminding me.  That's done now.  I also updated the description of the i_glob and j_glob implementation in the create_blocks subroutine.

File: cicecore/cicedyn/infrastructure/ice_domain.F90 [hunk 3]
```diff
@@ -259,6 +259,12 @@ subroutine init_domain_blocks
       call abort_ice(subname//' ERROR: Not enough ghost cells allocated', file=__FILE__, line=__LINE__)
    endif
 
+   if (ns_boundary_type == 'closed') then
+      call abort_ice(subname//' ERROR: ns_boundary_type = '//trim(ns_boundary_type)//' not supported', file=__FILE__, line=__LINE__)
+   endif
+   if (ew_boundary_type == 'closed') then
+      call abort_ice(subname//' ERROR: ew_boundary_type = '//trim(ew_boundary_type)//' not supported', file=__FILE__, line=__LINE__)
+   endif
```
[the remark this answers is not in the document]
Comment (TillRasmussen) [reply cicecore/cicedyn/infrastructure/ice_domain.F90]: I agree that the expected behavior of closed boundaries is that the ice hits a wall/land

**Review by eclare108213 — APPROVED** [review]
Approved, pending a few changes to comments/docs as noted.  Thank you!
